File: vision/smooth.js
// vision/detect.js
import cv from 'opencv4nodejs';
import { ROI, THRESH } from './config.js';
import { mapRoi } from './detect.js';
import { readGreenNumber } from './counters.js';
import { detectTrump } from './trump.js';
import { detectCardInSlot } from './cards.js';
import fs from 'node:fs';
import path from 'node:path';

function loadTmps(dir) {
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.png'));
  return files.map(f => ({ name: path.basename(f, '.png'), mat: cv.imdecode(fs.readFileSync(path.join(dir, f))) }));
}
const rankTmps = loadTmps('vision/templates/ranks');
const suitTmps = loadTmps('vision/templates/suits');

export async function detectFromScreenshot(pngBuffer, viewportW, viewportH) {
  const shot = cv.imdecode(pngBuffer); // BGR Mat
  const shotW = shot.cols, shotH = shot.rows;

  // Counters
  const deckRect = mapRoi(ROI.deckCount, shotW, shotH, viewportW, viewportH);
  const takenOppRect = mapRoi(ROI.takenOpp, shotW, shotH, viewportW, viewportH);
  const takenMeRect = mapRoi(ROI.takenMe, shotW, shotH, viewportW, viewportH);

  const deckCount = await readGreenNumber(shot.getRegion(deckRect));
  const takenOpp  = await readGreenNumber(shot.getRegion(takenOppRect));
  const takenMe   = await readGreenNumber(shot.getRegion(takenMeRect));

  // Trump
  const trumpRect = mapRoi(ROI.trumpSlot, shotW, shotH, viewportW, viewportH);
  const trump = detectTrump(shot.getRegion(trumpRect));

  // Hand slots
  const hand = [];
  for (const slot of ROI.handSlots) {
    const r = mapRoi(slot, shotW, shotH, viewportW, viewportH);
    const info = detectCardInSlot(shot.getRegion(r), rankTmps, suitTmps);
    hand.push(info.card); // can be null if uncertain this frame
  }

  return {
    deckCount, takenOpp, takenMe,
    trump: trump.rank && trump.suit ? `${trump.rank}-${trump.suit}` : null,
    hand
  };
}

File: vision/smooth.py
# vision/smooth.py
import cv2 as cv
import numpy as np
from .config import ROI
from .detect import map_roi
from .counters import read_green_number
from .trump import detect_trump
from .cards import detect_card_in_slot
from .templates import rank_templates, suit_templates


def detect_from_screenshot(png_buffer, viewport_w, viewport_h):
    shot = cv.imdecode(np.frombuffer(png_buffer, np.uint8), cv.IMREAD_COLOR)
    shot_h, shot_w = shot.shape[:2]

    deck_rect = map_roi(ROI['deckCount'], shot_w, shot_h, viewport_w, viewport_h)
    taken_opp_rect = map_roi(ROI['takenOpp'], shot_w, shot_h, viewport_w, viewport_h)
    taken_me_rect = map_roi(ROI['takenMe'], shot_w, shot_h, viewport_w, viewport_h)

    x, y, w, h = deck_rect
    deck_count = read_green_number(shot[y:y+h, x:x+w])
    x, y, w, h = taken_opp_rect
    taken_opp = read_green_number(shot[y:y+h, x:x+w])
    x, y, w, h = taken_me_rect
    taken_me = read_green_number(shot[y:y+h, x:x+w])

    trump_rect = map_roi(ROI['trumpSlot'], shot_w, shot_h, viewport_w, viewport_h)
    x, y, w, h = trump_rect
    trump_info = detect_trump(shot[y:y+h, x:x+w])

    hand = []
    for slot in ROI['handSlots']:
        r = map_roi(slot, shot_w, shot_h, viewport_w, viewport_h)
        x, y, w, h = r
        card_info = detect_card_in_slot(shot[y:y+h, x:x+w], rank_templates, suit_templates)
        hand.append(card_info['card'])

    trump_card = None
    if trump_info['rank'] and trump_info['suit']:
        trump_card = f"{trump_info['rank']}-{trump_info['suit']}"

    return {
        'deckCount': deck_count,
        'takenOpp': taken_opp,
        'takenMe': taken_me,
        'trump': trump_card,
        'hand': hand,
    }
